README.md
This repository contains plugin for the APT package tool which adds support for accessing authenticated Artifact Registry repositories.

debian/changelog
apt-transport-artifact-registry (1:20220518.00-g1ubuntu1) kinetic; urgency=medium

  * New release.

 -- Miriam España Acebal <miriam.espana@canonical.com>  Fri, 20 May 2022 12:41:31 +0200


apt/method_test.go
//  Copyright 2021 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package apt

import (
	"bufio"
	"context"
	"io"
	"net/http"
	"testing"
)

func TestHandleConfigure(t *testing.T) {
	var tests = []struct {
		configItems []string
		expected    aptMethodConfig
	}{
		{
			[]string{
				"Acquire::gar::Service-Account-JSON=/path/to/creds.json",
				"Acquire::gar::Service-Account-Email=email-address@domain",
			},
			aptMethodConfig{serviceAccountJSON: "/path/to/creds.json"},
		},
		{
			[]string{
				"Acquire::gar::Service-Account-Email=email-address@domain",
			},
			aptMethodConfig{serviceAccountEmail: "email-address@domain"},
		},
		{
			[]string{
				"some::other::config=value",
			},
			aptMethodConfig{},
		},
	}

	for _, tt := range tests {
		method := &AptMethod{config: &aptMethodConfig{}}
		msg := &AptMessage{
			code:        601,
			description: "Configuration",
			fields:      map[string][]string{"Config-Item": tt.configItems},
		}

		method.handleConfigure(msg)
		if method.config.serviceAccountJSON != tt.expected.serviceAccountJSON {
			t.Errorf("path config items don't match, got %q expected %q", method.config.serviceAccountJSON, tt.expected.serviceAccountJSON)
		}
		if method.config.serviceAccountEmail != tt.expected.serviceAccountEmail {
			t.Errorf("email config items don't match, got %q expected %q", method.config.serviceAccountEmail, tt.expected.serviceAccountEmail)
		}

	}

}

type fakeHTTPClient struct {
	code   int
	header map[string][]string
}

func (m fakeHTTPClient) Do(req *http.Request) (*http.Response, error) {
	if m.code == 0 {
		m.code = 200
	}
	if m.header == nil {
		m.header = map[string][]string{"Content-Length": []string{"200"}, "Last-Modified": []string{"whenever"}}
	}
	return &http.Response{StatusCode: m.code, Header: m.header}, nil
}

type fakeDownloader struct{}

func (d fakeDownloader) download(_ io.ReadCloser, _ string) (string, error) {
	return "ABCDEFGHI", nil
}

func TestAptMethodRun(t *testing.T) {

	stdinreader, stdinwriter := io.Pipe()
	stdoutreader, stdoutwriter := io.Pipe()
	workMethod := NewAptMethod(bufio.NewReader(stdinreader), stdoutwriter)
	workMethod.client = fakeHTTPClient{}
	workMethod.dl = fakeDownloader{}

	ctx := context.Background()
	ctx2, cancel := context.WithCancel(ctx)
	go workMethod.Run(ctx2)

	reader := AptMessageReader{reader: bufio.NewReader(stdoutreader)}
	msg, err := reader.ReadMessage(ctx)
	if err != nil {
		t.Errorf("failed, %v", err)
	}
	if msg.code != 100 || msg.description != "Capabilities" {
		t.Errorf("failed, didn't receive capabilities message")
	}

	writer := AptMessageWriter{writer: stdinwriter}
	writer.WriteMessage(AptMessage{
		code:        601,
		description: "Configuration",
		fields:      map[string][]string{"Config-Item": []string{"Acquire::gar::Service-Account-Email=email@domain"}},
	})

	writer.WriteMessage(AptMessage{
		code:        600,
		description: "URI Acquire",
		fields:      map[string][]string{"URI": []string{"http://fake.uri"}, "Filename": []string{"/path/to/file"}},
	})

	msg, err = reader.ReadMessage(ctx)
	if err != nil {
		t.Errorf("failed, %v", err)
	}
	if msg.code != 200 || msg.description != "URI Start" ||
		msg.Get("URI") != "http://fake.uri" || msg.Get("Size") != "200" {
		t.Errorf("failed, didn't receive uri start message. msg is %q", msg)
	}

	msg, err = reader.ReadMessage(ctx)
	if err != nil {
		t.Fatalf("failed, %v", err)
	}
	if msg.code != 201 || msg.description != "URI Done" ||
		msg.Get("URI") != "http://fake.uri" || msg.Get("Filename") != "/path/to/file" {
		t.Errorf("failed, didn't receive uri start message. msg is %q", msg)
	}
	cancel()

	// This was set after we sent the 601, but for guaranteed timing in the
	// test we don't check until after we've read a reply message.
	if workMethod.config.serviceAccountEmail != "email@domain" {
		t.Errorf("failed, didn't set method configuration. got %v, expected %q", workMethod.config.serviceAccountEmail, "email@domain")
	}

	for _, p := range []io.Closer{stdinreader, stdinwriter, stdoutreader, stdoutwriter} {
		if err := p.Close(); err != nil {
			t.Errorf("Error from %v: %v", p, err)
		}
	}
}

func TestAptMethodRun404(t *testing.T) {

	stdinreader, stdinwriter := io.Pipe()
	stdoutreader, stdoutwriter := io.Pipe()
	workMethod := NewAptMethod(bufio.NewReader(stdinreader), stdoutwriter)
	workMethod.client = fakeHTTPClient{code: 404}
	workMethod.dl = fakeDownloader{}

	ctx := context.Background()
	ctx2, cancel := context.WithCancel(ctx)
	defer cancel()
	go workMethod.Run(ctx2)

	reader := AptMessageReader{reader: bufio.NewReader(stdoutreader)}
	msg, err := reader.ReadMessage(ctx)
	if err != nil {
		t.Fatalf("failed, %v", err)
	}
	if msg.code != 100 || msg.description != "Capabilities" {
		t.Errorf("failed, didn't receive capabilities message")
	}

	writer := AptMessageWriter{writer: stdinwriter}
	writer.WriteMessage(AptMessage{
		code:        600,
		description: "URI Acquire",
		fields:      map[string][]string{"URI": []string{"http://fake.uri"}, "Filename": []string{"/path/to/file"}},
	})

	msg, err = reader.ReadMessage(ctx)
	if err != nil {
		t.Fatalf("failed, %v", err)
	}
	if msg.code != 400 || msg.description != "URI Failure" ||
		msg.Get("URI") != "http://fake.uri" || msg.Get("Message") == "" {
		t.Errorf("failed, didn't receive uri failure message. msg is %q", msg)
	}
	cancel()

	for _, p := range []io.Closer{stdinreader, stdinwriter, stdoutreader, stdoutwriter} {
		if err := p.Close(); err != nil {
			t.Errorf("Error from %v: %v", p, err)
		}
	}
}

func TestAptMethodRun304(t *testing.T) {

	stdinreader, stdinwriter := io.Pipe()
	stdoutreader, stdoutwriter := io.Pipe()
	workMethod := NewAptMethod(bufio.NewReader(stdinreader), stdoutwriter)
	workMethod.client = fakeHTTPClient{code: 304}
	workMethod.dl = fakeDownloader{}

	ctx := context.Background()
	ctx2, cancel := context.WithCancel(ctx)
	defer cancel()
	go workMethod.Run(ctx2)

	reader := AptMessageReader{reader: bufio.NewReader(stdoutreader)}
	msg, err := reader.ReadMessage(ctx)
	if err != nil {
		t.Fatalf("failed, %v", err)
	}
	if msg.code != 100 || msg.description != "Capabilities" {
		t.Errorf("failed, didn't receive capabilities message")
	}

	writer := AptMessageWriter{writer: stdinwriter}
	writer.WriteMessage(AptMessage{
		code:        600,
		description: "URI Acquire",
		fields:      map[string][]string{"URI": []string{"http://fake.uri"}, "Filename": []string{"/path/to/file"}},
	})

	msg, err = reader.ReadMessage(ctx)
	if err != nil {
		t.Fatalf("failed, %v", err)
	}
	if msg.code != 201 || msg.description != "URI Done" ||
		msg.Get("URI") != "http://fake.uri" || msg.Get("IMS-Hit") != "true" {
		t.Errorf("failed, didn't receive uri done message. msg is %q", msg)
	}
	cancel()

	for _, p := range []io.Closer{stdinreader, stdinwriter, stdoutreader, stdoutwriter} {
		if err := p.Close(); err != nil {
			t.Errorf("Error from %v: %v", p, err)
		}
	}
}

func TestAptMethodRunFail(t *testing.T) {

	stdinreader, stdinwriter := io.Pipe()
	stdoutreader, stdoutwriter := io.Pipe()
	workMethod := NewAptMethod(bufio.NewReader(stdinreader), stdoutwriter)
	workMethod.client = fakeHTTPClient{code: 404}
	workMethod.dl = fakeDownloader{}

	ctx := context.Background()
	ctx2, cancel := context.WithCancel(ctx)
	defer cancel()
	go workMethod.Run(ctx2)

	reader := AptMessageReader{reader: bufio.NewReader(stdoutreader)}
	msg, err := reader.ReadMessage(ctx)
	if err != nil {
		t.Fatalf("failed, %v", err)
	}
	if msg.code != 100 || msg.description != "Capabilities" {
		t.Errorf("failed, didn't receive capabilities message")
	}

	writer := AptMessageWriter{writer: stdinwriter}
	writer.WriteMessage(AptMessage{
		code:        700,
		description: "Bogus method",
	})

	msg, err = reader.ReadMessage(ctx)
	if err != nil {
		t.Fatalf("failed, %v", err)
	}
	if msg.code != 401 || msg.description != "General Failure" || msg.Get("Message") == "" {
		t.Errorf("failed, didn't receive general failure message. msg is %q", msg)
	}
	cancel()

	for _, p := range []io.Closer{stdinreader, stdinwriter, stdoutreader, stdoutwriter} {
		if err := p.Close(); err != nil {
			t.Errorf("Error from %v: %v", p, err)
		}
	}
}


apt/message_test.go
//  Copyright 2021 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package apt

import (
	"bufio"
	"bytes"
	"context"
	"testing"
)

func TestAptMessageGet(t *testing.T) {
	var tests = []struct {
		message  AptMessage
		expected string
	}{
		{
			// Happy case.
			AptMessage{code: 123, description: "Fake", fields: map[string][]string{"key": []string{"val1", "val2"}}},
			"val1",
		},
		{
			// Missing key.
			AptMessage{code: 123, description: "Fake", fields: map[string][]string{"some-other-key": []string{"val1", "val2"}}},
			"",
		},
		{
			// Missing value.
			AptMessage{code: 123, description: "Fake", fields: map[string][]string{"key": []string{}}},
			"",
		},
	}

	for _, tt := range tests {
		if res := tt.message.Get("key"); res != tt.expected {
			t.Errorf("failed, expected: %q got: %q", tt.expected, res)
		}
	}
}

func TestAptWriterWriteMessage(t *testing.T) {
	var tests = []struct {
		message  AptMessage
		expected string
	}{
		{
			AptMessage{
				code:        123,
				description: "Fake",
				fields: map[string][]string{
					"akey": []string{"val1", "val2"},
					"zkey": []string{"val4"},
					"Zkey": []string{"val3"},
				},
			},
			// Capital letters before lowercase, then alphabetical.
			"123 Fake\nZkey: val3\nakey: val1\nakey: val2\nzkey: val4\n\n",
		},
		{
			AptMessage{
				code: 123,
				// Missing description.
				fields: map[string][]string{
					"akey": []string{"val1"},
				},
			},
			"123 \nakey: val1\n\n",
		},
		{
			AptMessage{
				// missing code.
				description: "Fake",
				fields: map[string][]string{
					"akey": []string{"val1"},
				},
			},
			"0 Fake\nakey: val1\n\n",
		},
		{
			AptMessage{
				code:        123,
				description: "Fake",
				// missing fields.
			},
			"123 Fake\n\n",
		},
		{
			AptMessage{
				code:        123,
				description: "Fake",
				fields: map[string][]string{
					// Missing field value.
					"akey": []string{},
				},
			},
			"123 Fake\n\n",
		},
	}

	for _, tt := range tests {
		var buffer bytes.Buffer
		writer := NewAptMessageWriter(&buffer)
		err := writer.WriteMessage(tt.message)
		if err != nil || buffer.String() != tt.expected {
			t.Errorf("failed, expected:\n%q\ngot:\n%q", tt.expected, buffer.String())
		}
	}
}

func TestAptWriterSendCapabilities(t *testing.T) {
	var buffer bytes.Buffer
	writer := NewAptMessageWriter(&buffer)
	expected := "100 Capabilities\nSend-Config: true\nVersion: 1.0\n\n"
	if err := writer.SendCapabilities(); err != nil || buffer.String() != expected {
		t.Errorf("failed, expected:\n%q\ngot:\n%q", expected, buffer.String())
	}
}

func TestAptWriterLog(t *testing.T) {
	var tests = []struct {
		msg, expected string
	}{
		{
			"some log message",
			"101 Log\nMessage: some log message\n\n",
		},
	}

	for _, tt := range tests {
		var buffer bytes.Buffer
		writer := NewAptMessageWriter(&buffer)
		if err := writer.Log(tt.msg); err != nil || buffer.String() != tt.expected {
			t.Errorf("failed, expected:\n%q\ngot:\n%q", tt.expected, buffer.String())
		}
	}
}
func TestAptWriterURIStart(t *testing.T) {
	var tests = []struct {
		uri, size, lastModified, expected string
	}{
		{
			"http://fake.uri/debian/",
			"419304",
			"Mon, 01 Mar 2021 03:05:06 GMT",
			"200 URI Start\nLast-Modified: Mon, 01 Mar 2021 03:05:06 GMT\nResume-Point: 0\nSize: 419304\nURI: http://fake.uri/debian/\n\n",
		},
	}

	for _, tt := range tests {
		var buffer bytes.Buffer
		writer := NewAptMessageWriter(&buffer)
		if err := writer.URIStart(tt.uri, tt.size, tt.lastModified); err != nil || buffer.String() != tt.expected {
			t.Errorf("failed, expected:\n%q\ngot:\n%q", tt.expected, buffer.String())
		}
	}
}

// func URIDone(uri, size, lastModified, md5Hash, filename string, ims bool)
func TestAptWriterURIDone(t *testing.T) {
	var tests = []struct {
		uri, size, lastModified, md5Hash, filename, expected string
		ims                                                  bool
	}{
		{
			"http://fake.uri/debian/",
			"419304",
			"Mon, 01 Mar 2021 03:05:06 GMT",
			"ABCDEFGHIJKL",
			"/some/local/filename",
			"201 URI Done\nFilename: /some/local/filename\nLast-Modified: Mon, 01 Mar 2021 03:05:06 GMT\nMD5-Hash: ABCDEFGHIJKL\nSize: 419304\nURI: http://fake.uri/debian/\n\n",
			false,
		},
		{
			"http://fake.uri/debian/",
			"419304",
			"Mon, 01 Mar 2021 03:05:06 GMT",
			"ABCDEFGHIJKL",
			"/some/local/filename",
			"201 URI Done\nFilename: /some/local/filename\nIMS-Hit: true\nLast-Modified: Mon, 01 Mar 2021 03:05:06 GMT\nURI: http://fake.uri/debian/\n\n",
			true,
		},
	}

	for _, tt := range tests {
		var buffer bytes.Buffer
		writer := NewAptMessageWriter(&buffer)
		if err := writer.URIDone(tt.uri, tt.size, tt.lastModified, tt.md5Hash, tt.filename, tt.ims); err != nil || buffer.String() != tt.expected {
			t.Errorf("failed, expected:\n%q\ngot:\n%q", tt.expected, buffer.String())
		}
	}
}

func TestAptWriterFailURI(t *testing.T) {
	var tests = []struct {
		uri, msg, expected string
	}{
		{
			"http://fake.uri/debian/",
			"uri failure message",
			"400 URI Failure\nMessage: uri failure message\nURI: http://fake.uri/debian/\n\n",
		},
	}

	for _, tt := range tests {
		var buffer bytes.Buffer
		writer := NewAptMessageWriter(&buffer)
		if err := writer.FailURI(tt.uri, tt.msg); err != nil || buffer.String() != tt.expected {
			t.Errorf("failed, expected:\n%q\ngot:\n%q", tt.expected, buffer.String())
		}
	}
}

func TestAptWriterFail(t *testing.T) {
	var tests = []struct {
		msg, expected string
	}{
		{
			"general failure message",
			"401 General Failure\nMessage: general failure message\n\n",
		},
	}

	for _, tt := range tests {
		var buffer bytes.Buffer
		writer := NewAptMessageWriter(&buffer)
		if err := writer.Fail(tt.msg); err != nil || buffer.String() != tt.expected {
			t.Errorf("failed, expected:\n%q\ngot:\n%q", tt.expected, buffer.String())
		}
	}
}

func compareFields(first, second map[string][]string) bool {
	if len(first) != len(second) {
		return false
	}
	for key, firstVals := range first {
		secondVals, ok := second[key]
		if !ok || len(firstVals) != len(secondVals) {
			return false
		}
		for idx, firstVal := range firstVals {
			if firstVal != secondVals[idx] {
				return false
			}
		}
	}
	return true
}

func TestAptReaderReadMessage(t *testing.T) {
	var tests = []struct {
		msg      string
		expected AptMessage
	}{
		{
			"123 Fake Header\nField1: val1\nField2: val2\nField2: val3\nField1: val4\n\n",
			AptMessage{
				code:        123,
				description: "Fake Header",
				fields: map[string][]string{
					"Field1": []string{"val1", "val4"},
					"Field2": []string{"val2", "val3"},
				},
			},
		},
	}

	for _, tt := range tests {
		var buffer bytes.Buffer
		buffer.WriteString(tt.msg)
		reader := NewAptMessageReader(bufio.NewReader(&buffer))
		res, err := reader.ReadMessage(context.Background())
		if err != nil {
			t.Errorf("failed: %v", err)
		}
		if res.code != tt.expected.code ||
			res.description != tt.expected.description ||
			!compareFields(res.fields, tt.expected.fields) {
			t.Errorf("failed, expected: %v got: %v", tt.expected, res)
		}
	}
}

func TestAptReaderParseHeader(t *testing.T) {
	var tests = []struct {
		message  AptMessage
		header   string
		expected AptMessage
	}{
		{
			AptMessage{},
			"123 Fake Code",
			AptMessage{code: 123, description: "Fake Code"},
		},
	}

	for _, tt := range tests {
		reader := AptMessageReader{message: &AptMessage{}}
		if err := reader.parseHeader(tt.header); err != nil {
			t.Errorf("failed, %v", err)
		}
		if reader.message.code != tt.expected.code ||
			reader.message.description != tt.expected.description {
			t.Errorf("failed, expected: %v got: %v", tt.expected, reader.message)
		}
	}
}

func TestAptReaderParseHeaderFail(t *testing.T) {
	var tests = []struct {
		message AptMessage
		header  string
	}{
		{
			AptMessage{},
			"123FakeCode", // Invalid format.
		},
		{
			AptMessage{},
			"Xxx Fake Code", // Invalid format.
		},
		{
			AptMessage{},
			"", // Empty header.
		},
		{
			AptMessage{code: 123},
			"600 URI Acquire", // Valid header, existing message.
		},
		{
			AptMessage{description: "blah"},
			"600 URI Acquire", // Valid header, existing message.
		},
	}

	for idx, tt := range tests {
		reader := AptMessageReader{message: &tt.message}
		if err := reader.parseHeader(tt.header); err == nil {
			t.Errorf("validation failed test %d", idx)
		}
	}
}

func TestAptReaderParseField(t *testing.T) {
	var tests = []struct {
		message  AptMessage
		field    string
		expected map[string][]string
	}{
		{
			// Test initial fields.
			AptMessage{},
			"Field1: val1",
			map[string][]string{"Field1": []string{"val1"}},
		},
		{
			// Test appending fields.
			AptMessage{fields: map[string][]string{"Field1": []string{"val1"}}},
			"Field1: val2",
			map[string][]string{"Field1": []string{"val1", "val2"}},
		},
	}

	for _, tt := range tests {
		reader := AptMessageReader{message: &tt.message}
		if err := reader.parseField(tt.field); err != nil {
			t.Errorf("failed, %v", err)
		}
		if !compareFields(tt.message.fields, tt.expected) {
			t.Errorf("failed, expected: %v got %v", tt.expected, tt.message.fields)
		}
	}
}

func TestAptReaderParseFieldFail(t *testing.T) {
	var tests = []struct {
		field string
	}{
		{
			"Field1 = val1",
		},
		{
			"",
		},
		{
			":val1",
		},
		{
			" : val1",
		},
	}

	for idx, tt := range tests {
		reader := AptMessageReader{message: &AptMessage{}}
		if err := reader.parseField(tt.field); err == nil {
			t.Errorf("validation failed test %d", idx)
		}
	}
}


apt/method.go
//  Copyright 2021 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package apt

import (
	"bufio"
	"context"
	"crypto/md5"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"strings"

	"golang.org/x/oauth2"
	"golang.org/x/oauth2/google"
)

const (
	cloudPlatformScope = "https://www.googleapis.com/auth/cloud-platform"
)

// NewAptMethod returns an AptMethod.
func NewAptMethod(input *bufio.Reader, output io.Writer) *AptMethod {
	return &AptMethod{
		config: &aptMethodConfig{},
		writer: NewAptMessageWriter(output),
		reader: NewAptMessageReader(input),
		dl:     downloaderImpl{},
	}
}

// httpClient exists to enable mocking of http.Client.
type httpClient interface {
	Do(req *http.Request) (*http.Response, error)
}

// downloader exists to enable mocking of AptMethod.download.
type downloader interface {
	download(io.ReadCloser, string) (string, error)
}

type downloaderImpl struct{}

// AptMethod represents the method handler.
type AptMethod struct {
	reader *AptMessageReader
	writer *AptMessageWriter
	config *aptMethodConfig
	client httpClient
	dl     downloader
}

type aptMethodConfig struct {
	serviceAccountJSON, serviceAccountEmail string
}

// Run runs the method.
func (m *AptMethod) Run(ctx context.Context) {
	m.writer.SendCapabilities()
	for {
		select {
		case <-ctx.Done():
			return
		default:
		}
		msg, err := m.reader.ReadMessage(ctx)
		if err != nil {
			continue
		}
		switch msg.code {
		case 600:
			m.handleAcquire(msg)
		case 601:
			m.handleConfigure(msg)
		default:
			// TODO: now write a test for this.
			m.writer.Fail(fmt.Sprintf("Unsupported message code %d received from apt", msg.code))
		}
	}
}

func (m *AptMethod) initClient() error {
	if m.client != nil {
		return nil
	}

	var ts oauth2.TokenSource
	ctx := context.Background()
	switch {
	case m.config.serviceAccountJSON != "":
		json, err := ioutil.ReadFile(m.config.serviceAccountJSON)
		if err != nil {
			return fmt.Errorf("Failed to obtain creds: %v", err)
		}
		creds, err := google.CredentialsFromJSON(ctx, json, cloudPlatformScope)
		if err != nil {
			return fmt.Errorf("Failed to obtain creds: %v", err)
		}
		ts = creds.TokenSource
	case m.config.serviceAccountEmail != "":
		ts = google.ComputeTokenSource(m.config.serviceAccountEmail)
	default:
		creds, err := google.FindDefaultCredentials(ctx)
		if err != nil {
			return fmt.Errorf("Failed to obtain creds: %v", err)
		}
		ts = creds.TokenSource
	}
	if ts == nil {
		return errors.New("Failed to obtain creds")
	}
	m.client = oauth2.NewClient(ctx, ts)
	return nil
}

// download performs the actual downloading to target file and returns
// an MD5 hash of the downloaded file.
func (r downloaderImpl) download(body io.ReadCloser, filename string) (string, error) {
	defer body.Close()
	data, err := ioutil.ReadAll(body)
	if err != nil {
		return "", err
	}
	file, err := os.Create(filename)
	if err != nil {
		return "", err
	}
	defer file.Close()

	_, err = file.Write(data)
	return fmt.Sprintf("%x", md5.Sum(data)), err
}

func (m *AptMethod) handleAcquire(msg *AptMessage) error {
	uri := msg.Get("URI")
	if uri == "" {
		err := errors.New("No URI provided in Acquire message")
		m.writer.Fail(err.Error())
		return err
	}
	filename := msg.Get("Filename")
	if filename == "" {
		err := errors.New("No filename provided in Acquire message")
		m.writer.FailURI(uri, err.Error())
		return err
	}
	ifModifiedSince := msg.Get("Last-Modified")

	if err := m.initClient(); err != nil {
		m.writer.FailURI(uri, err.Error())
		return err
	}

	realuri := strings.Replace(uri, "ar+https", "https", 1)
	req, err := http.NewRequest("GET", realuri, nil)
	if err != nil {
		return err
	}
	if ifModifiedSince != "" {
		// TODO: validate this string is in RFC1123Z format.
		req.Header.Add("If-Modified-Since", ifModifiedSince)
	}
	resp, err := m.client.Do(req)
	if err != nil {
		m.writer.FailURI(uri, err.Error())
		return err
	}

	size := resp.Header.Get("Content-Length")
	lastModified := resp.Header.Get("Last-Modified")
	switch resp.StatusCode {
	case 200:
		// It's weird to send URI Start after we've already contacted
		// the server, but we need to know the size.
		m.writer.URIStart(uri, size, lastModified)
		md5Hash, err := m.dl.download(resp.Body, filename)
		if err != nil {
			m.writer.FailURI(uri, err.Error())
			return err
		}
		m.writer.URIDone(uri, size, lastModified, md5Hash, filename, false)
	case 304:
		// Unchanged since Last-Modified. Respond with "IMS-Hit: true" to
		// indicate the existing file is valid.
		m.writer.URIDone(uri, size, lastModified, "", filename, true)
	default:
		// All other codes including 404, 403, etc.
		err := fmt.Errorf("Error downloading: code %v", resp.StatusCode)
		m.writer.FailURI(uri, err.Error())
		return err
	}

	return nil
}

func (m *AptMethod) handleConfigure(msg *AptMessage) {
	configs, ok := msg.fields["Config-Item"]
	if !ok {
		// Nothing to set.
		return
	}
	for _, configItem := range configs {
		if strings.Contains(configItem, "Acquire::gar::Service-Account-JSON") {
			parts := strings.SplitN(configItem, "=", 2)
			if len(parts) != 2 {
				// TODO: log this?
				return
			}
			m.config.serviceAccountJSON = strings.TrimSpace(parts[1])
		}
		if strings.Contains(configItem, "Acquire::gar::Service-Account-Email") {
			parts := strings.SplitN(configItem, "=", 2)
			if len(parts) != 2 {
				// TODO: log this?
				return
			}
			m.config.serviceAccountEmail = strings.TrimSpace(parts[1])
		}
	}
	// Enforce the precedence of these two options.
	if m.config.serviceAccountJSON != "" {
		m.config.serviceAccountEmail = ""
	}
}


cmd/ar+https/main.go
//  Copyright 2021 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package main

import (
	"bufio"
	"context"
	"os"

	"github.com/GoogleCloudPlatform/artifact-registry-apt-transport/apt"
)

func main() {
	apt := apt.NewAptMethod(bufio.NewReader(os.Stdin), os.Stdout)
	apt.Run(context.Background())
}


apt/message_writer.go
//  Copyright 2021 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package apt

import (
	"io"
)

// AptMessageWriter supports writing Apt messages.
type AptMessageWriter struct {
	writer io.Writer
}

// NewAptMessageWriter returns an AptMessageWriter.
func NewAptMessageWriter(w io.Writer) *AptMessageWriter {
	return &AptMessageWriter{writer: w}
}

// WriteMessage writes an AptMessage.
func (w *AptMessageWriter) WriteMessage(m AptMessage) error {
	return w.writeString(m.String())
}

// WriteString writes a raw string.
func (w *AptMessageWriter) writeString(s string) error {
	if _, err := w.writer.Write([]byte(s)); err != nil {
		return err
	}
	return nil
}

// SendCapabilities writes a 100 Capabilities message.
func (w *AptMessageWriter) SendCapabilities() error {
	return w.WriteMessage(new100Message())
}

// Log writes a 101 Log message.
func (w *AptMessageWriter) Log(msg string) error {
	return w.WriteMessage(new101Message(msg))
}

// URIStart writes a 200 URI Start message.
func (w *AptMessageWriter) URIStart(uri, size, lastModified string) error {
	return w.WriteMessage(new200Message(uri, size, lastModified))
}

// URIDone writes a 201 URI Done message.
func (w *AptMessageWriter) URIDone(uri, size, lastModified, md5Hash, filename string, ims bool) error {
	return w.WriteMessage(new201Message(uri, size, lastModified, md5Hash, filename, ims))
}

// FailURI writes a 400 URI Failure message.
func (w *AptMessageWriter) FailURI(uri, msg string) error {
	return w.WriteMessage(new400Message(uri, msg))
}

// Fail writes a 401 General Failure message.
func (w *AptMessageWriter) Fail(msg string) error {
	return w.WriteMessage(new401Message(msg))
}


apt/message.go
//  Copyright 2021 Google LLC
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

package apt

import (
	"fmt"
	"sort"
	"strings"
)

// AptMessage represents a single RFC822 Apt message.
type AptMessage struct {
	code        int
	description string
	fields      map[string][]string
}

// Get returns the first AptMessage Field for `key`, or "".
func (m *AptMessage) Get(key string) string {
	if vals, ok := m.fields[key]; ok {
		if len(vals) > 0 {
			return vals[0]
		}
	}
	return ""
}

func (m *AptMessage) String() string {
	// Map iteration is unordered. For testing convenience, write alphabetical output.
	sortedKeys := make([]string, len(m.fields))
	i := 0
	for key := range m.fields {
		sortedKeys[i] = key
		i++
	}
	sort.Strings(sortedKeys)

	message := []string{fmt.Sprintf("%d %s", m.code, m.description)}
	for _, key := range sortedKeys {
		for _, val := range m.fields[key] {
			message = append(message, fmt.Sprintf("%s: %s", key, val))
		}
	}
	message = append(message, "")
	message = append(message, "") // End with a newline.
	return strings.Join(message, "\n")
}

func new100Message() AptMessage {
	fields := make(map[string][]string)
	fields["Send-Config"] = []string{"true"}
	fields["Version"] = []string{"1.0"}
	return AptMessage{code: 100, description: "Capabilities", fields: fields}
}

func new101Message(msg string) AptMessage {
	fields := make(map[string][]string)
	fields["Message"] = []string{msg}
	return AptMessage{code: 101, description: "Log", fields: fields}
}

func new200Message(uri, size, lastModified string) AptMessage {
	fields := make(map[string][]string)
	fields["URI"] = []string{uri}
	fields["Size"] = []string{size}
	if lastModified != "" {
		fields["Last-Modified"] = []string{lastModified}
	}
	fields["Resume-Point"] = []string{"0"}
	return AptMessage{code: 200, description: "URI Start", fields: fields}
}

func new201Message(uri, size, lastModified, md5Hash, filename string, imsHit bool) AptMessage {
	fields := make(map[string][]string)
	fields["URI"] = []string{uri}
	fields["Last-Modified"] = []string{lastModified}
	fields["Filename"] = []string{filename}
	if imsHit {
		fields["IMS-Hit"] = []string{"true"}
	} else {
		fields["Size"] = []string{size}
		fields["MD5-Hash"] = []string{md5Hash}
	}
	return AptMessage{code: 201, description: "URI Done", fields: fields}
}

func new400Message(uri, msg string) AptMessage {
	fields := make(map[string][]string)
	fields["URI"] = []string{uri}
	fields["Message"] = []string{msg}
	return AptMessage{code: 400, description: "URI Failure", fields: fields}
}

func new401Message(msg string) AptMessage {
	fields := make(map[string][]string)
	fields["Message"] = []string{msg}
	return AptMessage{code: 401, description: "General Failure", fields: fields}
}
